File: rgn_concurrency_akka_essentials/src/main/java/org/rgn/first/mapreduce/MapReduceApplication.java
package org.rgn.first.mapreduce;

import java.util.concurrent.TimeUnit;

import org.rgn.first.mapreduce.actors.MasterActor;
import org.rgn.first.mapreduce.messages.Result;

import scala.concurrent.Await;
import scala.concurrent.Future;
import scala.concurrent.duration.Duration;
import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.actor.Props;
import akka.pattern.Patterns;
import akka.util.Timeout;

public class MapReduceApplication {

	public static void main(String[] args) throws Exception {
		Timeout timeout = new Timeout(Duration.create(5, TimeUnit.SECONDS));
		ActorSystem _system = ActorSystem.create("MapReduceApp");

		ActorRef master = _system.actorOf(Props.create(MasterActor.class));
		master.tell(
				"The quick brownfox tried to jumo over the lazy dog and fell on the dog",
				null);

		master.tell("lazy dog and fell on the dog", null);
		master.tell("Dog is man's best friend", null);
		master.tell("Dog and Fox belong to the same family", null);

		Thread.sleep(5000);

		Future<Object> future = Patterns.ask(master, new Result(), timeout);

		String result = (String) Await.result(future, timeout.duration());

		System.out.println(result);
		_system.shutdown();
	}

}

File: rgn_concurrency/rgn_concurrency_akka_essentials/src/main/scala/org/rgn/scala/first/mapreduce/MapReduceApplication.scala
package org.rgn.scala.first.mapreduce

import scala.collection.mutable.ArrayBuffer
import scala.concurrent.duration._
import org.rgn.scala.first.mapreduce.actors.MasterActor
import akka.actor.ActorSystem
import akka.actor.Props
import akka.util.Timeout
import scala.concurrent.Await
import akka.pattern.ask

sealed trait MapReduceMessage
case class WordCount(word: String, count: Int) extends MapReduceMessage
case class MapData(dataList: ArrayBuffer[WordCount]) extends MapReduceMessage
case class ReduceData(reduceDataMap: Map[String, Int]) extends MapReduceMessage
case class Result() extends MapReduceMessage

object MapReduceApplication extends App {
  val _system = ActorSystem("MapReduceApp")
  val master = _system.actorOf(Props[MasterActor], name = "master")

  implicit val timeout = Timeout(5 seconds)

  master ! "The quick brown fox tried to jump over the lazy dog and fell on the dog"
  master ! "Dog is man's best friend"
  master ! "Dog and Fox belong to the same family"

  Thread.sleep(500)
  val future = (master ? Result).mapTo[String]
  val result = Await.result(future, timeout.duration)
  println(result)
  _system.shutdown

}